hedera-node/hedera-app/src/main/java/com/hedera/node/app/records/impl/BlockStreamManagerImpl.java

```diff
@@ -0,0 +1,400 @@
+        this.blockStreamProducer = requireNonNull(blockStreamProducer);
+
+        // Get static configuration that is assumed not to change while the node is running
+        final var blockStreamConfig = configProvider.getConfiguration().getConfigData(BlockStreamConfig.class);
+        this.numRoundsInBlock = blockStreamConfig.numRoundsInBlock();
+        if (this.numRoundsInBlock <= 0) {
+            throw new IllegalArgumentException("numRoundsInBlock must be greater than 0");
+        }
+        this.numBlockHashesToKeepBytes = blockStreamConfig.numOfBlockHashesInState() * HASH_SIZE;
+
+        // Initialize the last block info and provisional block info.
+        // NOTE: State migration happens BEFORE dagger initialization, and this object is managed by dagger. So we are
+        // guaranteed that the state exists PRIOR to this call.
+        final var states = state.getReadableStates(BlockRecordService.NAME);
+        final var blockInfoState = states.<BlockInfo>getSingleton(V0490BlockRecordSchema.BLOCK_INFO_STATE_KEY);
+        this.lastBlockInfo = blockInfoState.get();
+        assert this.lastBlockInfo != null : "Cannot be null, because this state is created at genesis";
+        this.blockOpen = false;
+        this.roundsUntilNextBlock = numRoundsInBlock;
+
+        // Initialize the stream file producer. NOTE, if the producer cannot be initialized, and a random exception is
+        // thrown here, then startup of the node will fail. This is the intended behavior. We MUST be able to produce
+        // block streams, or there really is no point to running the node!
+        final var runningHashState =
+                states.<RunningHashes>getSingleton(V0490BlockRecordSchema.RUNNING_HASHES_STATE_KEY);
+        final var lastRunningHashes = runningHashState.get();
+        assert lastRunningHashes != null : "Cannot be null, because this state is created at genesis";
+        this.blockStreamProducer.initFromLastBlock(lastRunningHashes, this.lastBlockInfo.lastBlockNumber());
+    }
+
+    @Override
+    public boolean startUserTransaction(
+            @NonNull Instant consensusTime, @NonNull HederaState state, @NonNull PlatformState platformState) {
+        if (EPOCH.equals(lastBlockInfo.firstConsTimeOfCurrentBlock())) {
+            // This is the first transaction of the first block, so set both the firstConsTimeOfCurrentBlock
+            // and the current consensus time to now
+            final var now = new Timestamp(consensusTime.getEpochSecond(), consensusTime.getNano());
+            lastBlockInfo = lastBlockInfo
+                    .copyBuilder()
+                    .consTimeOfLastHandledTxn(now)
+                    .firstConsTimeOfCurrentBlock(now)
+                    .build();
+            putLastBlockInfo(state);
+            return true;
+        }
+        return false;
+    }
+
+    @Override
+    public void advanceConsensusClock(@NonNull Instant consensusTime, @NonNull HederaState state) {
```

Comment:
This and a few other methods in this class are identical to `BlockRecordManagerImpl`. We may want to consider relocating the common code